### PR TITLE
fix(rome_cli): correctly compute file to handle in `rome ci`

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

There was an issue in the `rome ci` command, but we couldn't see it because the `MemoryFileSystem` behaves differently when handling the files, compared to `OsFileSystem`. By default, it was accepting any kind of file and it was using the `strip_prefix` function. This results in two issues:
- couldn't simulate commands like `rome format .` (the dot)
- couldn't replicate 100% how the real traversal behaves

This PR tries to change that behaviour.

## Test Plan

- I added a new test case for ignoring directories, and see if it works as expected.
- Once I added the `can_handle` function,  the test `ci_does_not_run_linter` was failing, because it was wrongly computing when a file is ignored. That's fixed in the `traversal.rs`

### CODE REVIEW

File: crates/rome_fs/src/fs/memory.rs [hunk 1]
```diff
@@ -161,10 +161,12 @@ impl<'scope> TraversalScope<'scope> for MemoryTraversalScope<'scope> {
         {
             let files = &self.fs.files.0.read();
             for path in files.keys() {
-                if path.strip_prefix(&base).is_ok() {
```
Comment (leops) [inline crates/rome_fs/src/fs/memory.rs:164]: That doesn't seem right, removing the `strip_prefix` check would cause all the files in the `MemoryFileSystem` to be visited as opposed to just the files under `base`, so calling `rome format --write src` would also reformat files that are not under `src/` (for instance in the `does_not_format_ignored_directories` test, the `scripts/test.js` file would get reformatted too)

Comment (ematipico) [reply crates/rome_fs/src/fs/memory.rs:164]: Using `strip_prefix` though, doesn't allow to run commands file `rome format ./`/`rome format .`. In fact, running this command would result in a `.is_ok()` to be `false`. What do you suggest to do?

Comment (leops) [reply crates/rome_fs/src/fs/memory.rs:164]: Unfortunately, correctly handling filesystem paths is inherently complex. Using `strip_prefix` would work correctly if those were absolute filesystem paths, but it becomes unreliable with relative paths. We can't simply resolve these to absolute paths with `canonicalize` since the files they point to do not exist in the underlying filesystem.
One solution would be to join the two relative paths to a root absolute path beforehand, eg. `let path = Path::new("/").join(path); let base = Path::new("/").join(base); path.strip_prefix(&base)` but I'm not sure that would work correctly on Windows.

Comment (ematipico) [reply crates/rome_fs/src/fs/memory.rs:164]: I applied your suggestion

File: crates/rome_fs/src/fs/memory.rs [hunk 2]
```diff
@@ -161,8 +161,21 @@ impl<'scope> TraversalScope<'scope> for MemoryTraversalScope<'scope> {
...
+                let should_process_file = if base.starts_with(".") || base.starts_with("./") {
```
Comment (leops) [inline crates/rome_fs/src/fs/memory.rs:164]: I think the `Path` type has a set of `is_relative` / `is_absolute` methods to detect these cases

File: crates/rome_fs/src/fs/memory.rs [hunk 3]
```diff
@@ -161,8 +161,21 @@ impl<'scope> TraversalScope<'scope> for MemoryTraversalScope<'scope> {
...
+                    // we simulate absolute paths
+                    let absolute_base = PathBuf::from("/");
```
Comment (leops) [inline crates/rome_fs/src/fs/memory.rs:166]: We should be joining `base` to this, otherwise `strip_prefix` will always return `Ok` even when the base doesn't actually match the path

**Review by leops — COMMENTED** [review]
Since the traversal behavior is hard to get right perhaps we should add more test cases to verify that it works correctly in various edge cases (check that it's visiting the correct files, but also importantly not visiting the files that are not contained in the base path)

File: crates/rome_fs/src/fs/memory.rs (hunk 2 above)
Comment (ematipico) [reply crates/rome_fs/src/fs/memory.rs:164]: That's not the case we are trying to get right. If we use `is_relative`, the paths `test.js` and `./` will both yield `true` in both cases. If we use `is_absolute`, it would give us `false` for both  paths.

**Comment by ematipico** [comment]
> Since the traversal behavior is hard to get right perhaps we should add more test cases to verify that it works correctly in various edge cases (check that it's visiting the correct files, but also importantly not visiting the files that are not contained in the base path)

I added another test to see if that works.

File: crates/rome_cli/tests/snap_test.rs [hunk 4]
```diff
@@ -138,13 +138,19 @@ impl From<SnapshotPayload<'_>> for CliSnapshot {
...
+        let mut files: Vec<_> = fs
+            .files()
+            .into_iter()
+            .map(|(file, entry)| {
+                let content = entry.lock();
+                let content = std::str::from_utf8(content.as_slice()).unwrap();
+                (file.to_str().unwrap().to_string(), String::from(content))
+            })
+            .collect();
+        files.sort_by_key(|(path, _)| path.to_string());
```
Comment (leops) [inline crates/rome_cli/tests/snap_test.rs:150]: Nit: there's no need to call `sort_by_key` here since the `BTreeMap` will re-sort the keys anyway

**Review by leops — APPROVED** [review]
